core/chains/evm/txmgr/finalizer.go

```diff
@@ -0,0 +1,275 @@
+					f.lggr.Errorw("Error processing head", "err", err)
+					f.SvcErrBuffer.Append(err)
+					continue
+				}
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (f *evmFinalizer) DeliverLatestHead(head *evmtypes.Head) bool {
+	return f.mb.Deliver(head)
+}
+
+func (f *evmFinalizer) ProcessHead(ctx context.Context, head *evmtypes.Head) error {
+	ctx, cancel := context.WithTimeout(ctx, processHeadTimeout)
+	defer cancel()
+	_, latestFinalizedHead, err := f.headTracker.LatestAndFinalizedBlock(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve latest finalized head: %w", err)
+	}
+	return f.processFinalizedHead(ctx, latestFinalizedHead)
+}
+
+// Determines if any confirmed transactions can be marked as finalized by comparing their receipts against the latest finalized block
+func (f *evmFinalizer) processFinalizedHead(ctx context.Context, latestFinalizedHead *evmtypes.Head) error {
+	// Cannot determine finality without a finalized head for comparison
+	if latestFinalizedHead == nil || !latestFinalizedHead.IsValid() {
+		return fmt.Errorf("invalid latestFinalizedHead")
+	}
+	earliestBlockNumInChain := latestFinalizedHead.EarliestHeadInChain().BlockNumber()
+	f.lggr.Debugw("processing latest finalized head", "block num", latestFinalizedHead.BlockNumber(), "block hash", latestFinalizedHead.BlockHash(), "earliest block num in chain", earliestBlockNumInChain)
+
+	// Retrieve all confirmed transactions, loaded with attempts and receipts
+	unfinalizedTxs, err := f.txStore.FindConfirmedTxes(ctx, f.chainId)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve confirmed transactions: %w", err)
+	}
+
+	var finalizedTxs []*Tx
+	// Group by block hash transactions whose receipts cannot be validated using the cached heads
+	receiptBlockHashToTx := make(map[common.Hash][]*Tx)
+	// Find transactions with receipt block nums older than the latest finalized block num and block hashes still in chain
+	for _, tx := range unfinalizedTxs {
+		receipt := tx.GetReceipt()
+		if receipt == nil || receipt.IsZero() || receipt.IsUnmined() {
+			f.lggr.AssumptionViolationw("invalid receipt found for confirmed transaction", "tx", tx, "receipt", receipt)
+			continue
+		}
+		// Receipt newer than latest finalized head block num
+		if receipt.GetBlockNumber().Cmp(big.NewInt(latestFinalizedHead.BlockNumber())) > 0 {
+			continue
+		}
+		// Receipt block num older than earliest head in chain. Validate hash using RPC call later
+		if receipt.GetBlockNumber().Int64() < earliestBlockNumInChain {
+			receiptBlockHashToTx[receipt.GetBlockHash()] = append(receiptBlockHashToTx[receipt.GetBlockHash()], tx)
+			continue
+		}
+		blockHashInChain := latestFinalizedHead.HashAtHeight(receipt.GetBlockNumber().Int64())
+		// Receipt block hash does not match the block hash in chain. Transaction has been re-org'd out but DB state has not been updated yet
+		if blockHashInChain.String() != receipt.GetBlockHash().String() {
+			continue
```

Comment:
There are some paths in this code, like this case, where we find out that a Confirmed Tx is actually not finalized, because the receipt is not valid, or not in canonical chain, or we can't check onchain.
These cases to me, look like things which the Confirmer should've spotted, and marked the Tx as UnConfirmed, but it didn't.

Do you think Finalizer should mark these txs as Unconfirmed?
Or otherwise, add some logic in Confirmer to handle detecting and fixing these txs.

Reply:
I believe the Confirmer does in its `EnsureConfirmedTransactionsInLongestChain` method. That's why I just skip over these in the Finalizer to give the Confirmer a chance to pick up on the re-org and mark the tx accordingly. Since this is a separate component running on a separate loop, I do these checks just in case the Finalizer picks up these tx before the Confirmer does. I wouldn't want to add the logic here to avoid both components sharing this re-org protection responsibility.